### PR TITLE
[CARBONDATA-1373] Enhance update performance by increasing parallelism

### PR DESCRIPTION
**Pull request by xuchuanyin**
# Scenario

Recently I have tested the update feature provided in Carbondata and found its poor performance.

I had a table containing about 14 million records with about 370 columns(no dictionary columns) and the data files are about 3.8 GB in total. All the data files were in one segment.

I performed an update SQL which update a column for all the records and the SQL looked like `UPDATE myTable SET (col1)=(col1+1000) WHERE TRUE`. In my environment, the update job failed with 'executor lost errors'. And I found 'spill data' related messages in the container logs.

# Analyze

I've read about the implementation of update-delete in Carbondata in ISSUE#440. The update consists a delete and an insert operation. And the error occurred during the insert operation.

After studying the code, I have found that while doing inserting, the updated records are grouped by the `segmentId`, which means all the recoreds in one segment will be processed in only one task, thus will cause task failure when the amount of input data is quite large.

# Solution

We should improve the parallelism when doing update for a segment.

~~I append a random key to the `segmentId` to increase the partition number before doing the insertion stage and then remove the suffix when doing the real insertion.~~

I customize a partitioner to distribute the records in one segment to multiple partitions, which will avoid a task to process to many records. And we use `partitionBy` to replace `groupByKey` to reduce shuffle overhead.

# Modification

+ Increase parallelism while processing one segment in update, this is achieved by distributing records to different partitions by using a customized partitioner.
+ Add a property to configure the parallelism.
+ Clean up local files after update (previous bugs)
+ Remove useless imports
+ Add tests
+ Add related documents

# Notes

I have tested in my example and the job finished in about 13 minutes successfully. The records were updated as expected.

Comparing to the previous implementation, the update performance has been enhanced:

Origin(Parallelism(1) + GroupBy): Update **FAILED**

Adding Parallelism(6) + GroupBy: Update **SUCCESSFULLY** using **13mins**

Parallelism(1) +  PartitionBy: Update **SUCCESSFULLY** using **21mins**

Adding Parallelism(6) + PartitionBy: Update **SUCCESSFULLY** using **5mins**

### CODE REVIEW

**Comment by xuchuanyin** [comment]
Previous PR #1253 is closed, use this new PR

**Comment by ravipesala** [comment]
SDV Build Failed with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/195/

**Comment by ravipesala** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/205/

**Comment by ravipesala** [comment]
SDV Build Failed with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/207/

**Comment by ravipesala** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/210/

**Comment by xuchuanyin** [comment]
Test failed for `DataMapWriterSuite.scala:123` with message:

List("blocklet start 0", "add page data: blocklet 0, page 0", "add page data: blocklet 0, page 1", "add page data: blocklet 0, page 2", "add page data: blocklet 0, page 3", "add page data: blocklet 0, page 4", "add page data: blocklet 0, page 5", "add page data: blocklet 0, page 6", "add page data: blocklet 0, page 7", "add page data: blocklet 0, page 8", "add page data: blocklet 0, page 9", "blocklet end: 0") did not equal List("blocklet start 0", "add page data: blocklet 0, page 0", "add page data: blocklet 0, page 1", "add page data: blocklet 0, page 2", "add page data: blocklet 0, page 3", "add page data: blocklet 0, page 4", "add page data: blocklet 0, page 5", "add page data: blocklet 0, page 6", "add page data: blocklet 0, page 7", "blocklet end: 0", "blocklet start 1", "add page data: blocklet 1, page 0", "add page data: blocklet 1, page 1", "blocklet end: 1")

which expected 2 blockets with 8 pages in total, and the actual result is 1 blocket with 8 pages in total.

Is this an occasional or real problem?

**Comment by xuchuanyin** [comment]
retest this

**Comment by xuchuanyin** [comment]
retest this please

**Comment by xuchuanyin** [comment]
@sraghunandan @jackylk pls review this PR.

Besides, please pay attention to the previous comment https://github.com/apache/carbondata/pull/1261#issuecomment-323000513, **there may be some errors in tests**.
 I just retest the PR without modifying the code, the test are passed.

**Comment by ravipesala** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/213/

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 1]
```diff
@@ -657,26 +652,53 @@ object CarbonDataRDDFactory {
...
-            nodes.distinct.toArray)
+        val partitionByRdd = keyRDD
+          .partitionBy(new SegmentPartitioner(segmentIdIndex, segmentUpdateParallelism))
 
-        res = groupBySegmentAndNodeRdd.map(x =>
-          triggerDataLoadForSegment(x._1, x._2.toIterator).toList
-        ).collect()
+        // because partitionId=segmentIdIndex*parallelism+RandomPart and RandomPart<parallelism,
+        // so segmentIdIndex=partitionId/parallelism, this has been verified.
+        res = partitionByRdd.map(_._2).mapPartitions(p => {
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: change `.mapPartitions(p => {` to `.mapPartitions{ partition =>`

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala [hunk 2]
```diff
@@ -756,26 +751,53 @@ object CarbonDataRDDFactory {
...
+            (row.get(row.size - 1).toString, Row(row.toSeq.slice(0, row.size - 1): _*)))
+
+        val loadMetadataDetails = SegmentStatusManager
+          .readLoadMetadata(carbonTable.getMetaDataFilepath)
+        val segmentIds = loadMetadataDetails.map(_.getLoadName)
+        val segmentIdIndex = segmentIds.zipWithIndex.toMap
+        val carbonTablePath = CarbonStorePath.getCarbonTablePath(carbonLoadModel.getStorePath,
+          carbonTable.getCarbonTableIdentifier)
+        val segmentId2maxTaskNo = segmentIds
+          .map(segId =>
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/carbondata/spark/rdd/CarbonDataRDDFactory.scala]: move to previous line, and change style to `.map{ segId =>`

**Comment by xuchuanyin** [comment]
@jackylk thks, reviews are fixed

**Comment by ravipesala** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/222/

**Comment by ravipesala** [comment]
SDV Build Success with Spark 2.1, Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/224/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**